### DISCUSSION TITLE
Should the 'row' role really be necessary for parents of 'gridcell' and other cell role elements?

### DISCUSSION THREAD

**Discussion by turjmner8** [w3c/aria#2128] [Ideas]
In general, would it be possible to remove the requirement for role 'gridcell' elements to be contained within role 'row' elements in a future spec? Newer DOM approaches like CSS grid are available to simplify the construction of grid structures that do not really support having a row-like element wrap all the given 'cell' elements of a logical row. Additionally, it's not quite clear to me how someone would be able to construct a 'row' role compliant grid for cases where cells span multiple rows without using TR elements, which introduces other sizing issues that do not exist when using more modern DIV / CSS strategies.

It seems like aria-rowindex, aria-colindex, etc... on 'gridcell' role elements themselves should be enough to convey what logical row and column a given 'gridcell' element belongs to even if its parent was a 'grid' role element rather than an additional 'row' role element.

I may be missing some nuance here, but the only thing that seems a little unclear would be not being able to set aria-selected on the containing 'row' element if row selection is supported. In general, it also feels like selections when dealing with tables/grids could also benefit from some enhancements though (how to tell if a row vs column vs individual cell is selected for example). Specifying aria-selected on 'row' role elements vs 'gridcell' role elements does not seem to make much of a difference in current screen reader impls (NVDA, JAWS, voiceover, etc...)

Thanks for any feedback / info here.

**Comment by MelSumner** [comment]
I got kind of nerd-sniped by this idea and made a codepen: https://codepen.io/melsumner/pen/BabgZmp

Does it accurately reflect what you're describing?

**Reply by turjmner8** [reply]
Yes, pretty much. Having aria-colcount and aria-rowcount on the outer 'grid' element, and each cell having aria-colindex and aria-rowindex seems like it should be enough without requiring a more stringent DOM structure.

If I use your example as a starting point, just adding that additional index information on each cell:
https://codepen.io/turjmner8/pen/abxbLYN

Specifically the 'row' element requirement seems to block the use of something like CSS grid from what otherwise could be a fully accessible 'grid' role implementation, and requires extra DOM elements that aren't required to present a visible grid to sighted users. That requirement really seems to be the thing conflating structure with style - in a large 2d grid, columns and rows and not inherently different, but the column context of cells is still available to screen readers even though there is no single column element those cells must be contained within. It just happens to feel more inline with the use of TR elements from earlier web development days.

**Comment by scottaohara** [comment]
related - https://github.com/w3c/aria/issues/1138 and my issue that we closed as duplicate https://github.com/w3c/aria/issues/1853

**Comment by nickygencs17** [comment]
+1 on this.
